File: 0257-binary-tree-paths/0257-binary-tree-paths.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
var binaryTreePaths = function(root) {
    const dfs = function (node, path) {
        if(!node.left && !node.right) {
            path.push(node.val);
            result.push(path.join('->'));
            path.pop();
            return;
        }

        path.push(node.val);
        node.left && dfs(node.left, path);
        node.right && dfs(node.right, path);
        path.pop();
    }

    const result = [];
    dfs(root, []);

    return result;
};

File: 0257-binary-tree-paths/0257-binary-tree-paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        result = []

        def dfs(node, path):
            if node.left == None and node.right == None:
                path.append(node.val)
                result.append("->".join(map(str, path)))
                path.pop()
                return
            
            path.append(node.val)
            if node.left != None:
                dfs(node.left, path)
            
            if node.right != None:
                dfs(node.right, path)

            path.pop()
        
        dfs(root, [])

        return result